server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -391,29 +431,131 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         return builder;
     }
 
+    public static Set<Role> BUILT_IN_ROLES = Set.of(MasterRole.INSTANCE, DataRole.INSTANCE, IngestRole.INSTANCE);
+    private static Map<String, Role> roleNameToPossibleRoles;
+
+    public static void setPossibleRoles(final Set<Role> possibleRoles) {
```

Comment:
I really don't like relying on setting static state here. It appears this is only used for local node construction? I wonder if the parsing of the roles could be moved out of here, into Node.LocalNodeFactory, and the createLocal method could take the set of roles. This way the LocalNodeFactory can take the gathered new roles from plugins.

Reply:
Alternatively, we could use SPI for gathering these roles, since there is zero state or setup needed for their construction.

Reply:
It’s needed in the `DiscoveryNode(StreamInput)` constructor too, that’s why I had to do this way as opposed to pushing through the `LocalNodeFactory`.

Reply:
What about just allowing a DiscoveryNodeRole to be constructed from the StreamInput? This means the node roles are not singletons, but I don't think that is a big deal. The bigger change would be the setting and how roles are mapped to settings since I do not believe we have a wire format for settings.

Reply:
I think the benefits of the strong-typing and association with the setting outweighs the ugliness of static state (and maybe can find a nice refactoring in the future?).

Reply:
I chatted with @rjernst; we will look into using SPI for this in a follow-up.